modules/packages/DistributedBagDeprecated.chpl

```diff
@@ -0,0 +1,1286 @@
+  it avoids the overhead of remotely accessing class fields, bounding performance
+  on communication.
+*/
+
+
+/* Implements a highly parallel segmented multiset.
+
+  .. warning::
+
+    The "DistributedBagDeprecated" module is deprecated and will be removed in a future
+    release. Please use the new "DistributedBag" module instead. If you have any concern
+    regarding this module, please let us know.
+
+  Summary
+  _______
+
+  A parallel-safe distributed multiset implementation that scales in terms of
+  nodes, processors per node (PPN), and workload; The more PPN, the more segments
+  we allocate to increase raw parallelism, and the larger the workload the better
+  locality (see :const:`distributedBagInitialBlockSize`). This data structure is unordered
+  and employs its own work stealing algorithm to balance work across nodes.
+
+  Usage
+  _____
+
+  To use :record:`DistBag`, the initializer must be invoked explicitly to
+  properly initialize the structure. Using the default state without initializing
+  will result in a halt.
+
+  .. code-block:: chapel
+
+    var bag = new DistBag(int, targetLocales=ourTargetLocales);
+
+
+  While the bag is safe to use in a distributed manner, each node always operates on it's privatized
+  instance. This means that it is easy to add data in bulk, expecting it to be distributed, when in
+  reality it is not; if another node needs data, it will steal work on-demand. This may not always be
+  desired, and likely will more memory consumption on a single node. We offer a way for the user to
+  invoke a more static load balancing approach, called :proc:`DistributedBagImpl.balance`, which will redistributed work.
+
+  .. code-block:: chapel
+
+    bag.addBulk(1..N);
+    bag.balance();
+
+  Possible Improvements
+  _____________________
+
+  1.  Dynamic work-stealing will require an overhaul to use a helper algorithm to keep down
+      the number of tasks spawned. Currently user tasks will wait on the current work-stealer
+      task, which will spawn is own helper tasks which act as shepherds, which then spawns more
+      in a fork-join fashion. This leads to an excessive amount of tasks being spawned at once.
+      To make matters worse, the waiting tasks don't even get any elements, nor does the work
+      stealing task, which opens up the possibility of live-lock where nodes steal work back
+      and forth before either can process it.
+  2.  Static work-stealing (A.K.A :proc:`DistributedBagImpl.balance`) requires a rework that performs a more distributed
+      and fast way of distributing memory, as currently 'excess' elements are shifted to a single
+      node to be redistributed in the next pass. On the note, we need to collapse the pass for moving
+      excess elements into a single pass, hopefully with a zero-copy overhead.
+
+  Methods
+  _______
+*/
+
+@deprecated("The DistributedBagDeprecated module is deprecated")
```

Comment:
I would still consider Michael the primary reviewer for this PR, but I saw the commit that added this file go past and wanted to point out that while deprecating and renaming a module at the same time will mean future uses of it are appropriately warned for, it does not alert current users that things have changed out from under them.

Since this module was not already considered stable, I think it is probably okay, but in the future I would recommend a slightly slower process of change, to ensure that users have warning and can opt-in to the new behavior instead of being surprised when things have changed.

Reply:
@lydia-duncan - that's something to take up with myself and @bradcray since @Guillaume-Helbecque here is just doing what I/we advised in https://github.com/chapel-lang/chapel/issues/21958#issuecomment-1819501798